security/fma-l1-upgrades.md

```diff
@@ -0,0 +1,329 @@
+
+#### **Mitigations:**
+
+1. In the first step to reset the initializer with varied layouts, a custom contract can be
+   created which hardcodes the exact modifications required for each contract with a unique
+   contracts specific function, such as `resetL1CrossDomainMessenger()`.
+2. Upgrading to use OpenZeppelin's unstructured storage for initialization state reduces the risk
+   of incorrectly reinitializing at the end of the upgrade, as it ensures that the `initializer`
+   modifier can operate on the same slot across contracts.
+3. After calling `OPCM.upgrade()` the `superchain-ops` script can use `vm.load()` and attempt to call `L1CrossDomainMessenger.upgrade()` (for example) to verify the intitialized status.
+4. Tenderly simulation should look for unexpected storage changes.
+
+#### **Detection:**
+
+Global event monitoring COULD be used on `monitorism` for the `Initialized` event, howevrer in the past similar monitoring
+has been deemed to noisy and not useful.
+
+#### **Recovery Path(s)**:
+
+An emergency upgrade would be required.
+
+### FM2: Storage Layout Corruption in an OP Chain Contract
+
+#### **Description:**
+
+During an upgrade, incorrect storage layout modifications could lead to
+corruption of contract state. This could happen if the upgrade process doesn't
+properly account for existing storage slots or if there are conflicts in the
+storage layout between versions.
+
+We disregard issues with the initializer as they are covered in FM1.
+
+#### **Risk Assessment:**
+
+High impact, Low likelihood
+
+#### **Mitigations:**
+
+Existing mitigations for this failure mode include:
+
+1. The use of spacer variables in L1 contracts to prevent storage collisions.
+2. The use of storage layout lock files tracked in git to detect breaking changes.
+3. The use of off-chain checks in the `superchain-ops` scripts to ensure that getters
+   continue to return the expected values.
+4. The use of state change validation prior to execution.
+
+Additionally, the [proposal](../protocol/l1-upgrades.md#modfications-to-l1-contracts) of adding a new minimally invasive
+`upgrade()` method to L1 contracts ensures minimal storage write operations during the upgrade process.
+
+#### **Detection:**
+
+Unexpected storage layout changes which are not caught by the mitigations above will be very difficult to detect, unless they
+immediately impact on the operation of the system. For example changing the value of `L1CrossDomainMessenger.otherMessenger()`
+will break both the messenger's `relayMessage()` and `sendMessage()` functions. Other values, such as an entry in a mapping or
+a rarely used address will be harder to detect.
+
+Regardless, strong emphasis should be placed on mitigation here.
+
+#### **Recovery Path(s)**:
+
+Emergency upgrade to fix the affected storage slots.
+
+### FM3: Storage Layout Corruption in the Upgrade Controller Safe
+
+#### **Description:**
+
+The decision was made to have the Upgrade Controller Safe DELEGATECALL
```

Comment:
An ugly but effective approach would be to record the most important Safe storage values at the start of the call then reset them afterwards

It looks like all of the OPCM state variables are in structs, which unfortunately solidity does not allow to be made `immutable`.

> I wonder if we enforced all state variables in opcm to be held at unstructured storage locations. This would probably be super ugly though.

[EIP-7201](https://eips.ethereum.org/EIPS/eip-7201) gives a much nicer syntax for accessing structs with unstructured storage. If we used that pattern, then it might be easier to detect when storage is being accessed. 

Though I do think that static analysis of the IR can be pretty robust.

Reply:
Make sense, resolving comment.